src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDirectoryDao.java

```diff
@@ -190,35 +193,64 @@ public void deleteDirectoryEntriesFromCollection(Firestore firestore, String col
+                            if (docSnap == null) {
+                                return null;
+                            }
+                            return docSnap.toObject(FireStoreDirectoryEntry.class);
+                        });
+
+                return fireStoreUtils.transactionGet("retrieveById", transaction);
+            } catch (Exception ex) {
+                if (FireStoreUtils.shouldRetry(ex)) {
+                    logger.warn("Retry-able error in firestore future get - file id: " +
+                        fileId + " message: " + ex.getMessage());
+                } else {
+                    throw new FileSystemExecutionException("get entry by id execution exception", ex);
+                }
+            }
 
-        return fireStoreUtils.transactionGet("retrieveById", transaction);
+            TimeUnit.MILLISECONDS.sleep(SLEEP_MILLISECONDS);
+        }
+        throw new FileSystemExecutionException("get entry by id failed - no more retries");
     }
 
     // Returns null if not found - upper layers do any throwing
     public FireStoreDirectoryEntry retrieveByPath(
         Firestore firestore, String collectionId, String fullPath) throws InterruptedException {
 
         String lookupPath = makeLookupPath(fullPath);
-        ApiFuture<FireStoreDirectoryEntry> transaction =
-            firestore.runTransaction(
-                xn -> {
-                    DocumentSnapshot docSnap = lookupByFilePath(firestore, collectionId, lookupPath, xn);
-                    if (docSnap == null) {
-                        return null;
-                    }
-                    return docSnap.toObject(FireStoreDirectoryEntry.class);
-                });
 
-        return fireStoreUtils.transactionGet("retrieveByPath", transaction);
+        for (int i = 0; i < RETRIES; i++) {
+            try {
+                ApiFuture<FireStoreDirectoryEntry> transaction =
+                    firestore.runTransaction(
+                        xn -> {
+                            DocumentSnapshot docSnap = lookupByFilePath(firestore, collectionId, lookupPath, xn);
+                            if (docSnap == null) {
+                                return null;
+                            }
+                            return docSnap.toObject(FireStoreDirectoryEntry.class);
+                        });
+
+                return fireStoreUtils.transactionGet("retrieveByPath", transaction);
+            } catch (Exception ex) {
+                if (FireStoreUtils.shouldRetry(ex)) {
+                    logger.warn("Retry-able error in firestore future get - path: " +
+                        lookupPath + " message: " + ex.getMessage());
+                } else {
+                    throw new FileSystemExecutionException("get entry by path execution exception", ex);
+                }
+            }
+
+            TimeUnit.MILLISECONDS.sleep(SLEEP_MILLISECONDS);
```

Comment:
I think you should be able to refactor this to avoid code duplication, as it looks like the function bodies are identical except for the lookup function and the message text for each. The only wrinkle is that the lookup function throws an exception, which means that you have to define an interface since you can't use `Function`. This is what I was able to come up with:

```
    interface LookupFunction {
        DocumentSnapshot apply(Transaction xn) throws InterruptedException;
    }

    // Returns null if not found - upper layers do any throwing
    private FireStoreDirectoryEntry runTransactionWithRetry(Firestore firestore,
                                                           LookupFunction function,
                                                           String transactionOp,
                                                           String warnMessage) throws InterruptedException {

        for (int i = 0; i < RETRIES; i++) {
            try {
                ApiFuture<FireStoreDirectoryEntry> transaction =
                    firestore.runTransaction(
                        xn -> {
                            DocumentSnapshot docSnap = function.apply(xn);
                            if (docSnap == null) {
                                return null;
                            }
                            return docSnap.toObject(FireStoreDirectoryEntry.class);
                        });

                return fireStoreUtils.transactionGet(transactionOp, transaction);
            } catch (Exception ex) {
                if (FireStoreUtils.shouldRetry(ex)) {
                    logger.warn("Retry-able error in firestore future get - " +
                        warnMessage + " message: " + ex.getMessage());
                } else {
                    throw new FileSystemExecutionException(transactionOp + " execution exception", ex);
                }
            }

            TimeUnit.MILLISECONDS.sleep(SLEEP_MILLISECONDS);
        }
        throw new FileSystemExecutionException(transactionOp + " failed - no more retries");

        Function
    }

    // Returns null if not found - upper layers do any throwing
    public FireStoreDirectoryEntry retrieveById(
        Firestore firestore, String collectionId, String fileId) throws InterruptedException {

        return runTransactionWithRetry(firestore, (xn) -> lookupByFileId(firestore, collectionId, fileId, xn),
            "retrieveById", " file id: " + fileId);
    }

    // Returns null if not found - upper layers do any throwing
    public FireStoreDirectoryEntry retrieveByPath(
        Firestore firestore, String collectionId, String fullPath) throws InterruptedException {

        String lookupPath = makeLookupPath(fullPath);

        return runTransactionWithRetry(firestore, (xn) -> lookupByFilePath(firestore, collectionId, lookupPath, xn),
            "retrieveByPath", " path: " + lookupPath);
    }
```